src/BuildLoop.cpp

```diff
@@ -245,18 +232,56 @@ static void AttemptCacheWrite(BuildQueue* queue, ThreadState* thread_state, Runt
+        if (forceReadTimestampFromDisk)
+            StatCacheMarkDirty(queue->m_Config.m_StatCache, non_generated_input_file.m_Filename, non_generated_input_file.m_FilenameHash);
+
+        uint64_t timeStamp = StatCacheStat(queue->m_Config.m_StatCache, non_generated_input_file.m_Filename, non_generated_input_file.m_FilenameHash).m_Timestamp;
+        if (latestTimestampSeenForNonGeneratedInputFile && timeStamp > *latestTimestampSeenForNonGeneratedInputFile)
+            *latestTimestampSeenForNonGeneratedInputFile = timeStamp;
+
+        HashAddInteger(&hashState, timeStamp);
+    }
+
+    HashDigest result;
+    HashFinalize(&hashState, &result);
+    return result;
+};
+
 static NodeBuildResult::Enum ExecuteNode(BuildQueue* queue, RuntimeNode* node, Mutex *queue_lock, ThreadState* thread_state, StatCache* stat_cache, const Frozen::DagDerived* dagDerived)
 {
     CheckDoesNotHaveLock(&queue->m_Lock);
 
-    if (IsNodeCacheableByLeafInputsAndCachingEnabled(queue,node))
-    {
-        CHECK(node->m_CurrentLeafInputSignature != nullptr);
-        bool stillTheSame = node->m_BuiltNode && node->m_BuiltNode->m_WasBuiltSuccessfully && node->m_BuiltNode->m_LeafInputSignature == node->m_CurrentLeafInputSignature->digest;
+    // Compute timestamps and record the latests file modification date we find for any input file not generated by the graph.
+    uint64_t latestTimestampSeenForNonGeneratedInputFile = 0;
+    auto timestampsHashOfNonGeneratedInputFiles = HashTimestampsOfNonGeneratedInputFiles(queue, node, false, &latestTimestampSeenForNonGeneratedInputFile);
 
-        if (!stillTheSame)
-            if (!VerifyAllVersionedFilesIncludedByGeneratedHeaderFilesWereAlreadyPartOfTheLeafInputs(queue, thread_state, node, dagDerived))
-                return NodeBuildResult::kRanFailed;
+    // Check latest file timestamp against filesystem "now"
+    bool thereIsAtLeastOneInputFileDatedInTheFuture = false;
+    if (latestTimestampSeenForNonGeneratedInputFile >= FileSystem::g_LastSeenFileSystemTime)
+    {
+        // Make sure we are in sync with current file system "now"
+        auto fileSystemTimeNow = FileSystemUpdateLastSeenFileSystemTime();
+        if (latestTimestampSeenForNonGeneratedInputFile == fileSystemTimeNow)
+        {
+            // If the latest modification time of any non generated input file matches now, then we wait for the next file system mtime tick.
+            // The reason we do this is so we can safely detect changes done by the user either during graph execution or in between
+            // two tundra executions happening within the same file system mtime frame.
+            FileSystemWaitUntilFileModificationDateIsInThePast(latestTimestampSeenForNonGeneratedInputFile);
+        }
+        else if (latestTimestampSeenForNonGeneratedInputFile > fileSystemTimeNow)
+        {
+            // If any file not generated by the graph is dated in the future we can't wait. Or we don't want to wait, since that could potentially
```

Comment:
Should we print a warning with the filename if we encounter a strange timestamp?

Reply:
This is normally not a big issue other than the node being rebuilt for each tundra invocation until it has a date in the past. And since we did added support for `bee why` I think that is good enough?

Reply:
`bee why` requires human intervention, whereas if we log something then it would appear in build logs in CI etc.

Reply:
but for a single build it doesn't really matter if you have timestamps in the future. it only matters if you are also concurrently modifying files or on the next build when you wonder why the same node was rebuilt again.

Reply:
also, we do log this info in the structured log (for bee why), only we don't log it to the console

Reply:
I guess my argument is that there are plenty of situations we could log to the console as to why a node is built but we normally don't do that. You could argue this is a special case, and infrequent enough to warrant a warning, but I'm not sure.

Reply:
>If we want to be extra helpful, then tundra could also print something at the end of the build (where it's visible to users) which mentions that there were wierd timestamps.

I don't think that is such a bad idea in general. To have an exit message type of thing where we print information about the build the user may be interested in. Probably only as very general statements, similar to what we do when we start the build.

Reply:
This idea has come up once or twice recently. Filed https://jira.unity3d.com/browse/DS-1357 so we can move forward with this PR.